File: architecture-skills/design-patterns/examples/template-method/template-method.java
// Template Method pattern in Java.
//
// The classic form: an abstract class with a final template method and
// abstract steps (compare java.io.InputStream or JUnit lifecycle hooks).
// The top-level class is non-public because "template-method" is not a
// valid Java class name.

import java.util.List;

class TemplateMethodDemo {

    record LineItem(String product, int units, double price) {
        double subtotal() {
            return units * price;
        }
    }

    abstract static class SalesReport {
        /** The template method: final, so the skeleton cannot vary. */
        final String generate(List<LineItem> items) {
            StringBuilder out = new StringBuilder(header()).append("\n");
            double total = 0;
            for (LineItem item : items) {
                out.append(row(item)).append("\n");
                total += item.subtotal();
            }
            return out.append(footer(total)).toString();
        }

        abstract String header();

        abstract String row(LineItem item);

        abstract String footer(double total);
    }

    static final class CsvReport extends SalesReport {
        @Override
        String header() {
            return "product,units,price,subtotal";
        }

        @Override
        String row(LineItem item) {
            return "%s,%d,%.2f,%.2f".formatted(item.product(), item.units(), item.price(), item.subtotal());
        }

        @Override
        String footer(double total) {
            return "total,,,%.2f".formatted(total);
        }
    }

    static final class MarkdownReport extends SalesReport {
        @Override
        String header() {
            return "| Product | Units | Subtotal |\n| --- | --- | --- |";
        }

        @Override
        String row(LineItem item) {
            return "| %s | %d | $%.2f |".formatted(item.product(), item.units(), item.subtotal());
        }

        @Override
        String footer(double total) {
            return "\n**Total: $%.2f**".formatted(total);
        }
    }

    public static void main(String[] args) {
        List<LineItem> items = List.of(
                new LineItem("Keyboard", 2, 80.00),
                new LineItem("Mouse", 1, 25.50),
                new LineItem("Monitor", 3, 210.00));

        for (SalesReport report : List.of(new CsvReport(), new MarkdownReport())) {
            System.out.println("--- " + report.getClass().getSimpleName() + " ---");
            System.out.println(report.generate(items));
            System.out.println();
        }
    }
}

File: architecture-skills/design-patterns/examples/template-method/template-method.cs
// Template Method pattern in C#.
//
// An abstract class with a non-virtual template method and abstract
// steps (compare ASP.NET's middleware/lifecycle hooks). C# code often
// prefers composition — injecting Func<> steps — over inheritance; the
// classic inheritance form is shown here.

using System;
using System.Collections.Generic;
using System.Text;

public sealed record LineItem(string Product, int Units, double Price)
{
    public double Subtotal => Units * Price;
}

public abstract class SalesReport
{
    // The template method: non-virtual, so the skeleton cannot vary.
    public string Generate(IReadOnlyList<LineItem> items)
    {
        var outText = new StringBuilder(Header()).Append('\n');
        double total = 0;
        foreach (var item in items)
        {
            outText.Append(Row(item)).Append('\n');
            total += item.Subtotal;
        }
        return outText.Append(Footer(total)).ToString();
    }

    protected abstract string Header();
    protected abstract string Row(LineItem item);
    protected abstract string Footer(double total);
}

public sealed class CsvReport : SalesReport
{
    protected override string Header() => "product,units,price,subtotal";

    protected override string Row(LineItem item) =>
        $"{item.Product},{item.Units},{item.Price:F2},{item.Subtotal:F2}";

    protected override string Footer(double total) => $"total,,,{total:F2}";
}

public sealed class MarkdownReport : SalesReport
{
    protected override string Header() =>
        "| Product | Units | Subtotal |\n| --- | --- | --- |";

    protected override string Row(LineItem item) =>
        $"| {item.Product} | {item.Units} | ${item.Subtotal:F2} |";

    protected override string Footer(double total) => $"\n**Total: ${total:F2}**";
}

public static class Program
{
    public static void Main()
    {
        var items = new List<LineItem>
        {
            new("Keyboard", 2, 80.00),
            new("Mouse", 1, 25.50),
            new("Monitor", 3, 210.00),
        };

        foreach (SalesReport report in new SalesReport[] { new CsvReport(), new MarkdownReport() })
        {
            Console.WriteLine($"--- {report.GetType().Name} ---");
            Console.WriteLine(report.Generate(items));
            Console.WriteLine();
        }
    }
}
